diff --git a/src/GestureManager.cpp b/src/GestureManager.cpp
index 777d9e1..b7f32f1 100644
--- a/src/GestureManager.cpp
+++ b/src/GestureManager.cpp
@@ -11,10 +11,9 @@
 #include <hyprland/src/config/ConfigValue.hpp>
 #include <hyprland/src/config/legacy/ConfigManager.hpp>
 #include <hyprland/src/desktop/state/FocusState.hpp>
-#include <hyprland/src/output/Monitor.hpp>
-#include <hyprland/src/pointer/PointerController.hpp>
-#include <hyprland/src/state/MonitorState.hpp>
+#include <hyprland/src/helpers/Monitor.hpp>
 #include <hyprland/src/managers/SeatManager.hpp>
+#include <hyprland/src/managers/PointerManager.hpp>
 #include <hyprland/src/managers/input/InputManager.hpp>
 #include <hyprland/src/managers/input/UnifiedWorkspaceSwipeGesture.hpp>
 #include <hyprland/src/plugins/PluginSystem.hpp>
@@ -359,7 +358,7 @@ void GestureManager::dragGestureUpdate(const wf::touch::gesture_event_t& ev) {
 
         case GestureType::LONG_PRESS: {
             const auto pos = this->m_sGestureState.get_center().current;
-            Pointer::pointerController()->warpTo(Vector2D(pos.x, pos.y));
+            g_pPointerManager->warpTo(Vector2D(pos.x, pos.y));
             g_pInputManager->simulateMouseMovement();
             return;
         }
@@ -590,8 +589,8 @@ void GestureManager::sendCancelEventsToWindows() {
 bool GestureManager::onTouchDown(ITouch::SDownEvent ev) {
     static auto const SEND_CANCEL = g_config->sendCancel;
 
-    auto monitor = State::monitorState()->query().name(!ev.device->m_boundOutput.empty() ? ev.device->m_boundOutput : "").run();
-    monitor      = monitor ? monitor : Desktop::focusState()->monitor();
+    auto monitor = !ev.device->m_boundOutput.empty() ? g_pCompositor->getMonitorFromName(ev.device->m_boundOutput) : nullptr;
+    monitor      = monitor ? monitor : Desktop::focusState()->monitor();
 
     if (!monitor) {
         Log::logger->log(Log::ERR, "[hyprgrass] onTouchDown: could not find a monitor???");
@@ -604,7 +603,7 @@ bool GestureManager::onTouchDown(ITouch::SDownEvent ev) {
     const auto& monitorSize = this->m_lastTouchedMonitor->m_size;
     this->m_monitorArea     = SMonitorArea{monitorPos.x, monitorPos.y, monitorSize.x, monitorSize.y};
 
-    Pointer::pointerController()->warpTo(
+    g_pPointerManager->warpTo(
         Vector2D{
             monitorPos.x + ev.pos.x * monitorSize.x,
             monitorPos.y + ev.pos.y * monitorSize.y,
